locksmith/src/controllers/hookController.ts

```diff
@@ -0,0 +1,131 @@
+// eslint-disable-next-line import/no-unresolved
+import { Response } from 'express-serve-static-core'
+import { networks } from '@unlock-protocol/networks'
+import * as z from 'zod'
+import fetch from 'cross-fetch'
+import crypto from 'crypto'
+import logger from '../logger'
+import { SignedRequest } from '../types'
+import { Hook } from '../models'
+
+const Hub = z.object({
+  topic: z.string().url(),
+  callback: z.string().url(),
+  mode: z.enum(['subscribe', 'unsubscribe']),
+  lease_seconds: z.number().optional(),
+  secret: z.string().optional(),
+})
+
+export async function subscribe(
+  hub: z.infer<typeof Hub>,
+  params: Record<string, string>
+) {
+  const hook = await Hook.findOne({
+    where: {
+      topic: hub.topic,
+      callback: hub.callback,
+    },
+  })
+
+  if (!hook) {
+    const check = isValidHubIntent(hub)
+    if (!check) {
+      const intentError = new Error(
+        `Subscriber failed to confirm intent for callback: ${hub.callback} on topic: ${hub.topic}.`
+      )
+      logger.error(intentError)
+      throw intentError
+    }
+
+    const expiration = new Date()
+    expiration.setSeconds(
+      expiration.getSeconds() + (hub.lease_seconds ?? 60 * 60 * 24 * 30)
+    )
+
+    const createdHook = await Hook.create({
+      expiration,
+      topic: hub.topic,
+      callback: hub.callback,
+      mode: hub.mode,
+      secret: hub.secret,
+      network: params.network,
+      lock: params.lock,
+    })
+
+    return createdHook
+  }
+
+  // If changing mode, reconfirm the intent and update
+  if (hook?.mode !== hub.mode) {
+    hook.mode = hub.mode
+    const check = isValidHubIntent(hub)
+    if (!check) {
+      logger.error(
+        `Subscriber failed to confirm intent for callback: ${hub.callback} on topic: ${hub.topic}.`
+      )
+      return
+    }
+    const updatedHook = await hook.save()
+    return updatedHook
+  }
+
+  return hook
```

Comment:
Should we add a case for when a hook already exists but is just extended?

Reply:
In that case, we should also reconfirm intent BTW, and change the secret if needed.

Reply:
Maybe I will just let people update all values instead of just mode :thinking: instead.

Reply:
Let's remain compliant with websub please :)

Reply:
okay, will add a case for increasing expiration date.

Reply:
Actually, let's leave this for later addition.

Reply:
Hum. Please make sure you keep track of it?